### ISSUE TITLE
feat request: Allow custom client path 

### ISSUE THREAD

**Issue by jackvane** [47ng/prisma-field-encryption#18]
I'm currently using two different database clients that live in a different directory from _@prisma/lcient_ as a workaround to support multiple databases as mentioned here:

https://github.com/prisma/prisma/issues/2443#issuecomment-1110434344

I'd like to be able to do something like this

```typescript
    const prismacli1 = new PrismaClient1();
    const prismacli2 = new PrismaClient2();

    prismacli1.$use(fieldEncryptionMiddleware());
    prismacli2.$use(fieldEncryptionMiddleware());
```

But is looks like the middleware explicitly injects into _@prisma/client_

```
Error: Cannot find module '@prisma/client'
Require stack:
- /Users/jangu/Development/bitrebel/bitrebel.io/api/node_modules/prisma-field-encryption/dist/dmmf.js
- /Users/jangu/Development/bitrebel/bitrebel.io/api/node_modules/prisma-field-encryption/dist/index.js
- /Users/jangu/Development/bitrebel/bitrebel.io/api/src/plugins/brapp.ts
- /Users/jangu/Development/bitrebel/bitrebel.io/api/src/index.ts
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/Users/jangu/Development/bitrebel/bitrebel.io/api/node_modules/prisma-field-encryption/dist/dmmf.js:4:18)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Module._compile (/Users/jangu/Development/bitrebel/bitrebel.io/api/node_modules/source-map-support/source-map-support.js:568:25)
    at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
    at Object.nodeDevHook [as .js] (/Users/jangu/Development/bitrebel/bitrebel.io/api/node_modules/ts-node-dev/lib/hook.js:63:13)
    at Module.load (node:internal/modules/cjs/loader:981:32)
```

It would be nice to be able to pass in an option to specify the location of my prisma clients 👍

**Comment by franky47** [comment]
That could be a nice improvement indeed.

Would you like to open a PR?

**Comment by synyster-jack** [comment]
You might be able to get to it faster than I could. I'm a bit stacked at the moment to implement the fix for it but from a glance at the source, here are my thoughts.

So ideally the user wants to pass the path of the client to the configuration and let the middleware use the DMMF from the supplied client.

I thought about allowing the path of the client in the config input and then dynamically loading the Prisma instance from there

_src/types.ts - l14_
```typescript
export interface Configuration {
  encryptionKey?: string
  decryptionKeys?: string[]
  clientPath?: string
}
```

_src/index.ts - l12_
```typescript
  let models: DMMFModels
  ;(async function () {
    if (config.clientPath) {
      const { Prisma } = await import(config.clientPath) // problem: this is untyped
      models = analyseDMMF(Prisma.dmmf)
    } else {
      models = analyseDMMF()
    }
  })()
```

Unfortunately, there are a few spots where this might be challenging though since you're sourcing the DMMF from the @prisma/client directory, particularly in the types and errors file. You'll need a clean way to source types from the custom client instead.

Also, this may not work clean without async/await because of the dynamic import.

I thought about maybe using module augmentation to extend DMMF types, but it wouldn't work with multiple clients.

**Comment by franky47** [comment]
> You might be able to get to it faster than I could. I'm a bit stacked at the moment to implement the fix for it but from a glance at the source, here are my thoughts.

Unfortunately I don't have a lot of free time to dedicate to this project at the moment.

> So ideally the user wants to pass the path of the client to the configuration and let the middleware use the DMMF from the supplied client.

That seems like the right option indeed.

Just so we're clear on your setup, are you using the same Prisma schema with two different client output locations, or are we talking about two different Prisma schemas (and associated clients) ?

> Unfortunately, there are a few spots where this might be challenging though since you're sourcing the DMMF from the @prisma/client directory, particularly in the types and errors file. You'll need a clean way to source types from the custom client instead.

Indeed, and it's not just about runtime code (where your fix could work), we're also relying on Prisma's generated TypeScript types for the DMMF. Though this could be something where generics could help, it would require more boilerplate in the default case (when using the default client located under `@prisma/client`) to pass it the right types and runtime objects.

**Comment by jackvane** [comment]
> Just so we're clear on your setup, are you using the same Prisma schema with two different client output locations, or are we talking about two different Prisma schemas (and associated clients) ?

I'm using two different clients that use two different schemas.

To be honest, implementing this feature without rearchitecting the whole library might not be feasible or even worth it. If Prisma somehow adds native [support for for multiple databases](https://github.com/prisma/prisma/issues/2443) that generates the stubs in _@prisma/client_, then you might have to refactor it again.

Thanks for your feedback though

**Comment by franky47** [comment]
@jackvane @synyster-jack @pwdevbcf: I have released [`v1.4.0-beta.1`](https://github.com/47ng/prisma-field-encryption/releases/tag/v1.4.0-beta.1) which adds support for custom/multiple Prisma clients, could you give it a try and tell me if it solves your respective use-cases please?

You'll need to pass the DMMF to the middleware rather than the path to where to find it:
```diff
  import { fieldEncryptionMiddleware } from 'prisma-field-encryption'
+ import { Prisma } from '../my/prisma/client/location'

  prismaClient.$use(
    fieldEncryptionMiddleware({
+    dmmf: Prisma.dmmf
    })
  )
```

Passing the DMMF directly opens up a lot of future possibilities in manipulating it at runtime, as opposed to a simple import path.

**Comment by pwdevbcf** [comment]
@franky47   thanks! This is working, there is only thing related to types, dmmf need only datamodel property, but dmmf from Prisma namespace has more than only datamodel, but this is not that big issue.

**Comment by franky47** [comment]
Does it give you a TypeScript error? My integration code seems to be OK with passing a fully fledged DMMF from Prisma to the internal DMMFDocument type that only declares what it uses.

**Comment by pwdevbcf** [comment]
```
(property) Configuration.dmmf?: {
    datamodel: {
        models: {
            name: string;
            fields: {
                name: string;
                isList: boolean;
                isUnique: boolean;
                isId: boolean;
                type: string;
                documentation?: string;
            }[];
        }[];
    };
}
Type 'Document' is not assignable to type '{ datamodel: { models: { name: string; fields: { name: string; isList: boolean; isUnique: boolean; isId: boolean; type: string; documentation?: string; }[]; }[]; }; }'.
  The types of 'datamodel.models' are incompatible between these types.
    Type 'Model[]' is not assignable to type '{ name: string; fields: { name: string; isList: boolean; isUnique: boolean; isId: boolean; type: string; documentation?: string; }[]; }[]'.
      Type 'Model' is not assignable to type '{ name: string; fields: { name: string; isList: boolean; isUnique: boolean; isId: boolean; type: string; documentation?: string; }[]; }'.
        Types of property 'fields' are incompatible.
          Type 'Field[]' is not assignable to type '{ name: string; isList: boolean; isUnique: boolean; isId: boolean; type: string; documentation?: string; }[]'.
            Type 'Field' is not assignable to type '{ name: string; isList: boolean; isUnique: boolean; isId: boolean; type: string; documentation?: string; }'.
              Types of property 'type' are incompatible.
                Type 'string | SchemaEnum | OutputType | SchemaArg' is not assignable to type 'string'.
                  Type 'SchemaEnum' is not assignable to type 'string'.ts(2322)
```

**Comment by franky47** [comment]
Can you share your TypeScript configuration (`tsconfig.json`), and what type of database are you using with this Prisma client?

**Comment by pwdevbcf** [comment]
Yep. tsconfig looks like:
```
{
   "compilerOptions": {
    "module": "commonjs",
    "esModuleInterop": true,
    "target": "ES2020",
    "noImplicitAny": false,
    "allowSyntheticDefaultImports": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "sourceMap": true,
    "outDir": "dist",
    "baseUrl": ".",
    "noEmit": false,
    "skipLibCheck": true
  },
  "ts-node": {
    "files": true
  },
  "exclude": ["test", "jest.config.js"],
}
```

and DB is PostgreSQL and `"@prisma/client": "^3.8.0"`

**Comment by pwdevbcf** [comment]
there is the same issue. I am not able to share schema - sorry. But thank you for the fast release! It helped me a lot

**Comment by franky47** [comment]
@jackvane @synyster-jack @pwdevbcf You may have noticed that Prisma released v4.0 last week, which includes a breaking change in their internals (the way we access the DMMF, which was not covered by semver anyway).

I'm thinking of releasing this feature (tracked in PR #20) with a peer dependency requirement on v4, in an upcoming major upgrade. Because the package where the DMMF lives has changed names, it would be hard to provide compatibility for both Prisma 3 and Prisma 4 in a single version.

This is how we'd go about accessing the DMMF, in the case of a default client location:

```ts
import { getDmmf } from '@prisma/internals'
import { fieldEncryptionMiddleware } from 'prisma-field-encryption'

prismaClient.$use(
  fieldEncryptionMiddleware({ dmmf: getDmmf() })
)
```

Now I don't know how this may translate to custom client locations, and how/if Prisma has had this use-case in mind in their package rearchitecturing. Are any of you planning to migrate (or already have) to Prisma 4?

**Comment by synyster-jack** [comment]
I have already migrated over to Prisma v4 so a peer dependency on it in a future release is fine with me.

**Comment by franky47** [comment]
How does it work to access the DMMF of a custom client then on v4?

**Comment by badyan** [comment]
This helped me a lot, confirm it works with Prisma 4.2.1.

**Comment by franky47** [comment]
Closing as implemented in `1.4.0` (for some reason the auto-close tag didn't pick this one up).

**Comment by subvertallchris** [comment]
Hey, just dropping a comment on this old issue to say THANK YOU for to everyone who contributed to this issue and the `dmmf` improvement back in 2022. I'm on day four of an exhausting migration to pnpm workspaces and this was the very last snag we hit. There was a moment of despair when I thought it could not be resolved but this did the trick.
